DOC: Tighten howto-docs guide

bjnath opened the pull request:
When generosity strikes, we want the user on board before they know
what happened to them, so the how-to should be as swift
and straightforward as we can make it. The existing page
has accomplished the crucial job of rounding up the facts;
this PR suggests how we might tighten and clarify the presentation.

<!-- Please be sure you are following the instructions in the dev guidelines
http://www.numpy.org/devdocs/dev/development_workflow.html
-->

<!-- We'd appreciate it if your commit message is properly formatted
http://www.numpy.org/devdocs/dev/development_workflow.html#writing-the-commit-message
-->

Commit:
ENH: improve printing of arrays with multi-line reprs

Commit:
Improve testing

Commit:
Improve way of breaking lines

Commit:
Make requested changes

WarrenWeckesser committed:
DEP: Deprecate `numpy.dual`.

Add a deprecation warning in the `numpy.dual` module, and
remove the use of `numpy.dual` from the few places where it
is used in the numpy code.

WarrenWeckesser committed:
DOC: dual: Note the deprecation of numpy.dual in the dual module docstring.

Also make the correction "Scipy" -> "SciPy" in a few places.

WarrenWeckesser committed:
REL: Add a release note about the deprecation of numpy.dual.

Commit:
Add release note snippet

seberg committed:
MAINT: Unify cached (C-level static) imports

The main idea is that once we unify them, we can think about actually
being able to also clean them up.
This is far away from having no global state, but at this time I do not
see that as a seriousl goal personally. But it could be possible
to reload the module safely in a single interpreter maybe.

The commit adds the cache import for the complex casts, since it doubles
performance for small casts when the warning is not suppressed and it
unifies the handling.

seberg committed:
BUG: Fix default fallback in genfromtxt

This affected (for example?) if the `dtype=object` was used
without a converter, meaning that the default one is used.
And this is currently the last one, which is `string_` (and thus
bytes).

Closes gh-16189

seberg committed:
TST: Add a test for np.add_docstring

Its not quite the right file, but close to newdoc seemed sensible
and we do not have a "right" file right now...

WarrenWeckesser committed:
DOC: Update the f2py section of the "Using Python as Glue" page.

* Update the output shown for the docstrings generated by f2py.

bjnath committed:
DOC: Tighten howto-docs guide

When generosity strikes, we want the user on board before they know
what happened to them, so the how-to should be as swift
and straightforward as we can make it. The existing page
has accomplished the crucial job of rounding up the facts;
this PR suggests how we might tighten and clarify the presentation.

bjnath committed:
DOC: Tighten howto-docs guide

When generosity strikes, we want the user on board before they know
what happened to them, so the how-to should be as swift
and straightforward as we can make it. The existing page
has accomplished the crucial job of rounding up the facts;
this PR suggests how we might tighten and clarify the presentation.

rlintott committed:
BUG: numpy.einsum indexing arrays now accept numpy int type (gh-16080)

* Using PyArray_PyIntAsIntp helper function instead

* TST: add tests for einsum numpy int and bool list subscripts
Added tests to check that einsum accepts numpy int64 types and
rejects bool. Rejecting bools is new behaviour in subscript lists.
I changed ValueError to TypeError on line 2496 in multiarraymodule.c
as it is more appropriate. I also modified einsumfunc.py to have the
same behaviour as in the C file when checking subscript list.
(Reject bools but accept anything else from operator.index())

Closes gh-15961

sibasmarak committed:
DOC: Updated the f2py section of the "Using Python as Glue" page. (#16258)

* Replace print statements with print function calls.
* Fix the filename in one of the commands: it is "add.f", not "add.f95".

seberg committed:
BUG: Add missing decref in fromarray error path

This function steals the dtype, so it must decref it also
on error. This is fixup of ffe76ac8df65c8e7831df9924a782276e060f3e6
which accidentally deleted the decref.

rossbar reviewed (commented):
Thanks for taking the time to look over this document (which in itself is relatively new)!

This is essentially a re-write of the document so I haven't looked over it in detail, but just wanted to draw attention to a couple points from the original version that were removed here:

 - I think the introduction section that lays out what documentation *is* is valuable - I think part of the inspiration for this document (and things like NEP 44) is to disentangle the many different forms of documentation that exist for the NumPy project. I think the original intro blurb did a nice job of this in a succinct form

 - The addition of documentation change "priorities" near the end of the document was intentional: changes to the documentation often require quite a bit of review, so the listing of "priorities" is intended to give users some guidance on what sorts of changes will have the greatest impact. It depends on the context, but there are often cases where changes that are arguably subjective (i.e. proposals for different wording but don't really change the semantics) end up soaking up quite a bit of attention. Maybe the inclusion of such guidelines is inappropriate here (we certainly don't want to discourage potential contributors!) but I just wanted to point out that it was intentional --- happy to hear what others think!

I'll circle back around to a more careful reading of all the changes when I get a chance - thanks again!

bjnath commented:
Thank you for comments!

If I asked you how to contribute and you replied by defining what documentation is, it wouldn't be a good conversation.  The reader has expectations that need to be addressed. The faster you assure them they're on the right page, the likelier they are to read on. 

> disentangle the many different forms of documentation that exist for the NumPy project

But NumPy isn't mentioned anywhere. Even if the paragraph began "NumPy documentation exists in a welter of forms," it would still seem out of place on a How to Contribute page.  The reader wonders:  "How does any of this help me to contribute?"

If the goal is to share ways to contribute documentation apart from writing a page, it would be great to add a paragraph addressing that. You've already listed the many forms that documentation can take; the new paragraph would also tell the reader how to contribute to each.
<br>

> documentation change "priorities" near the end

The section is extremely useful, and the PR moves it up to "First steps".  The points made by the original continue to ring out:

   > Fixes to technical inaccuracies take high priority...they are straightforward to review.
   > Changes to improve clarity are less straightforward.

charris committed:
ENH: Add tool for downloading release wheels from Anaconda.

This is a simplified version of terryfy::wheel-uploader that has
two advantages over the original:

- It works with Anaconda where our wheels are now stored.
- It is simplified to match the NumPY workflow.

rossbar commented:
> The section is extremely useful, and the PR moves it up to "First steps". The points made by the original continue to ring out:

Got it, thanks for pointing that it - unfortunately restructuring/moving text is one of those things that's very easy to miss in diffs (even the rich diffs don't help much). Thanks!

> But NumPy isn't mentioned anywhere. Even if the paragraph began "NumPy documentation exists in a welter of forms," it would still seem out of place on a How to Contribute page. The reader wonders: "How does any of this help me to contribute?"

Good point, it may be that the document conflates "how to" and "introductory" concepts, so tightening the focus would definitely be an improvement.

jbrockmendel committed:
ENH: include dt64/td64 isinstance checks in __init__.pxd

charris committed:
Merge pull request #16263 from seberg/decref-fromarray-error-path-fixup

BUG: Add missing decref in fromarray error path

mattip committed:
Merge pull request #16265 from charris/add-download-wheel-tool

ENH: Add tool for downloading release wheels from Anaconda.

tbm committed:
DOC: Fix typos and cosmetic issues

mattip committed:
Merge pull request #16269 from tbm/typo

DOC: Fix typos and cosmetic issues

charris committed:
REL: Update master after 1.19.x branch.

- Delete release note fragments from 1.19.0 (towncrier)
- Update 1.19.x release note (towncrier)
- Create 1.20.0-notes.rst for master development
- Update C-API versions for 1.20.x
- Update setup.py for 1.20.0

eric-wieser committed:
MAINT: Remove a pointless if

The `add` ufunc is happy to handle `out=None` by itself

eric-wieser committed:
MAINT: Avoid moving axes around multiple times

It's easier to move the relevant axis to position 0 in `ap` first than it is to move it for every relevant object simultaneously.

eric-wieser committed:
MAINT: Extract a lerp helper function to make the algorithm of quantile clearer

This does not affect the behavior in any way

eric-wieser committed:
MAINT: Remove special cases for 0d arrays in quantile

This also simplifies the axis handling logic, taking advantage of the fact we know `axis` is 0 for the rest of the function body

eric-wieser committed:
BUG: Ensure out argument is returned by identity for 0d arrays

This makes the following functions consistent with the behavior of ufuncs:

* `np.argmax`
* `np.argmin`
* `np.choose`
* `np.trace`
* `np.round`
* `np.take`

Previously they would unpack these into scalars.

charris committed:
Merge pull request #16275 from charris/update-after-1.19.x-branch

REL: Update master after 1.19.x branch.

SacredArrow committed:
DOC: Clarifications for np.std (#16267)

* DOC: Clarifications for np.std

* Line lengths fixes in np.std

* Swapped variable names in np.std

SacredArrow committed:
DOC: Clarifications for np.var

eric-wieser committed:
Merge pull request #16274 from eric-wieser/extract-lerp

MAINT: cleanups to quantile

WarrenWeckesser committed:
DOC: A few small changes in the "Using Python as Glue" page.

* Fix whitespace in the inputs to Python in the f2py examples.
* Capitalize Python and Fortran consistently.

mattip committed:
Merge pull request #16238 from seberg/maint-cached-imports

MAINT: Unify cached (C-level static) imports

dependabot-preview[bot] committed:
MAINT: Bump hypothesis from 5.12.0 to 5.14.0

Bumps [hypothesis](https://github.com/HypothesisWorks/hypothesis) from 5.12.0 to 5.14.0.
- [Release notes](https://github.com/HypothesisWorks/hypothesis/releases)
- [Commits](https://github.com/HypothesisWorks/hypothesis/compare/hypothesis-python-5.12.0...hypothesis-python-5.14.0)

Signed-off-by: dependabot-preview[bot] <support@dependabot.com>

melissawm commented:
Thanks for taking a look at this, @bjnath! Just to clarify: do you think the finer points that were removed really don't belong in the NumPy docs, or do you think they could be moved to a separate page somehow?

bjnath commented:
@melissawm -- By "finer points", do you mean the issues raised in @rossbar's two bullets?  

I think Ross and I concur that the "priorities" material still says what it said before.

And I think we agree as well that the original opening, which describes the forms documentation can take, is good writing that just needs the right home.

seberg committed:
Merge pull request #16257 from WarrenWeckesser/doc-glue-f2py

DOC: Update the f2py section of the "Using Python as Glue" page.

seberg committed:
Merge pull request #16278 from SacredArrow/var-doc-edit

DOC: Clarifications for ``np.var``.

seberg committed:
Merge pull request #16285 from numpy/dependabot/pip/hypothesis-5.14.0

MAINT: Bump hypothesis from 5.12.0 to 5.14.0

rk-mlu committed:
DOC: Calrify tiny/xmin in finfo and machar (gh-16253)

This commit adds a note to the documentation of finfo clarifying that
the attribute `tiny` does not represent the smallest positive usable
number. Instead it refers to the smallest positive number with there
being no leading 0's in the mantisse (normalized floating point number).
Alternatively, `tiny` represents the smallest positive floating point
number with full precision. The smallest positive subnormal floating
point number is typically orders of magnitudes smaller, but has reduced
precision since leading 0's in the mantisse are used to allow smaller
exponents than indicated by `minexp`.

The commit also adds a brief clarification to the docstring of machar.

Closes #16252

Co-authored-by: Anirudh Subramanian <anirudh2290@apache.org>
Co-authored-by: Ross Barnowski <rossbar@berkeley.edu>

guilhermeleobas committed:
DOC: Add PyArray_ContiguousFromObject C docs (gh-16207)

Add documentation for `PyArray_ContiguousFromObject` similar to the other
macros wrapping `PyArray_FromAny`

Fixes #16196

charris committed:
BUG: Fix tools/download-wheels.py.

`tools/download-wheels` was downloading an html page rather than the binary
wheel file.

mattip committed:
Merge pull request #16292 from charris/fix-download-wheels

BUG: Fix tools/download-wheels.py.

jbrockmendel committed:
Re-order declarations

jbrockmendel committed:
Missing declarations

jbrockmendel committed:
move below int64_t declaration

jbrockmendel committed:
troubleshoot CI

lkdmttg7 committed:
MAINT: Chain exceptions in generate_umath.py (gh-16254)

Co-authored-by: Sebastian Berg <sebastian@sipsolutions.net>
Co-authored-by: Eric Wieser <wieser.eric@gmail.com>

seberg committed:
Update numpy/lib/_iotools.py

Co-authored-by: Eric Wieser <wieser.eric@gmail.com>

charris committed:
BUG: Require Python >= 3.6 in setup.py

Closes #16294.

seberg committed:
Merge pull request #16276 from eric-wieser/take-out

 BUG: Ensure out argument is returned by identity for 0d arrays

bje- committed:
DOC: Add math.isclose to See Also (#16283)

* DOC: Move math.isclose note to See Also.

mattip committed:
Merge pull request #16295 from charris/fix-missed-3.5

BUG: Require Python >= 3.6 in setup.py

mattip committed:
Merge pull request #16242 from seberg/genfromtxt-default-regression

BUG: Fix default fallback in genfromtxt

eric-wieser committed:
BUG: np.info does not show keyword-only arguments

Using inspect.signature instead of `np.compat.getargspec` solves this problem.

`inspect.signature` also handles stripping the `self` argument of methods for us.

eric-wieser committed:
MAINT: remove workaround for fixed issue

take now correctly returns `out`, even on 0d arrays

seberg committed:
Merge pull request #16156 from WarrenWeckesser/deprecate-dual

DEP: Deprecate `numpy.dual`.

Commit:
Move the snippet note

blackpastures committed:
MAINT: precompute log(2.0 * M_PI) in `random_loggam' (gh-16237)

Most compilers should optimize it, but it doesn't hurt to inline and has a better name
now.

seberg committed:
BUG: Allow attaching documentation twice in add_docstring

This is technically not a bug, but some IDEs and IPython have
autoreload magic which can mean that NumPy gets reloaded a second
time. This is not safe, but when it happens ignoring that an
identical docstring is already attached fixes the issue.

seberg committed:
DOC: Fix generic scalar docstrings to point to ndarray

Previously some of these were not set, because they were already
set on the C-level. Remove the duplicate message and replace it
instead with a forward to the ``ndarray`` attribute/method.

These docstrings are inherited by the actual scalars and thus the
"virtual class" note was misleading.

seberg committed:
MAINT: Simplify logic in add_docstring

The old pointers are now provided by the C-API and the macros
seemed a bit confusing since ``new`` appears from nowhere being
defined in the macro.

Carreau committed:
DOC: Fix documentation rendering,

the => seem to be extra and rst seem to not interprete backtick with
leading space as closing a directive, the the full line on text was put
in a html pre-tag when building the docs.

see current https://numpy.org/doc/stable/reference/generated/numpy.packbits.html

seberg committed:
Merge pull request #16305 from Carreau/doc-micro-fix

DOC: Fix packbits documentation rendering,

seberg committed:
Merge pull request #16284 from eric-wieser/extract-lerp

MAINT: Clean up the implementation of quantile

mattip committed:
BUG: relpath fails for different drives on windows (#16308)

* BUG: relpath fails for different drives on windows

* ENH: always use abspath

Commit:
Improve the snippet

embray committed:
BLD: Avoid "visibility attribute not supported" warning (gh-16288)

I found that when building the latest master branch on Cygwin, while testing #16246, that thousands of warnings were generated at build time like:

numpy/core/src/npysort/binsearch.c.src: In function ‘binsearch_left_bool’:
numpy/core/src/npysort/binsearch.c.src:82:1: warning: visibility attribute not supported in this configuration; ignored [-Wattributes]
Granted this is just a warning, so I don't think it's a serious issue.

It seems the test that was supposed to check for __attribute__ support was not working as expected. The #pragmas only take effect if I provide a function body--they are ignored for bare declarations. I don't know if that's by intent, or if it's a GCC issue. For reference:

$ gcc --version
gcc (GCC) 7.4.0

panpiort8 committed:
Update doc/release/upcoming_changes/15997.improvement.rst

Co-authored-by: Eric Wieser <wieser.eric@gmail.com>

mattip committed:
Merge pull request #16239 from seberg/reimport-do-not-set-docs

BUG,DOC: Allow attach docs twice but error if wrong

mattip committed:
Merge pull request #15997 from panpiort8/array_pretty_print

ENH: improve printing of arrays with multi-line reprs

rossbar committed:
MAINT: Bumpy numpydoc version.

Fixes problem with citation (reference) anchors in rendered docs.

Actual fix was in numpy/numpydoc@7c42883, but this bumps to latest
(unreleased) numpydoc version.

seberg committed:
MAINT: Stop Using PyEval_Call* and simplify some uses

Python 3.9 started to deprecate these calls and the PyObject_Call*
variants are equivalent or nicer in any case.

seberg committed:
TST: Add 3.9-dev to the travis matrix

Co-authored-by: Charles Harris <charlesr.harris@gmail.com>

charris committed:
Merge pull request #16316 from rossbar/maint/bump_numpydoc

MAINT: Bump numpydoc version

charris committed:
Merge pull request #16318 from seberg/maint-python39

MAINT: Stop Using PyEval_Call* and simplify some uses

tinaoberoi committed:
DOC: Correct documentation of ``__array__`` when used as output array. (#16130)

lschroefl committed:
DOC: link np.interp to SciPy's multidimensional interpolation functions (#14154)

brandtbucher committed:
BUG: Don't segfault on bad __len__ when assigning. (gh-16327)

When __getitem__ fails, assignment falls back on __iter__, which may not
have the same length as __len__, resulting in a segfault.
See gh-7264.

* BUG: Don't rely on __len__ for safe iteration.

Skip __len__ checking entirely, since this has no guarantees of being
correct. Instead, first convert to PySequence_Fast and use that length.
Also fixes a refleak when creating a tmp array fails.
See gh-7264.

* TST: Update test for related edge-case.

The previous fix more gracefully handles this edge case by skipping the
__len__ check. Rather than raising with an unhelpful error message, just
create the array with whatever elements C() actually yields.
See gh-7264.

kevmo committed:
DOC: Fix spelling typo - homogenous to homogeneous. (#16324)

jbrockmendel committed:
Fix np.dtype declaration

lschroefl committed:
Update numpy/lib/function_base.py

Co-authored-by: Ross Barnowski <rossbar@berkeley.edu>

seberg committed:
BUG: Fix refcounting in add_newdoc

this is a trivial refcounting fix, since this function is mostly
used only at startup, the small refcounting issue was not noticed
until the new tests were added.

jbrockmendel committed:
First stab at tests, try on the CI

seberg committed:
CI: Create a link for the circleCI artifact

mattip committed:
Merge pull request #16332 from LSchroefl/link_SciPy_multidimensional_interpolation

DOC: link np.interp to SciPy's interpolation functions (closes #14154)

mattip committed:
ENH: run tests for __init__.pxd

rgommers committed:
Merge pull request #16333 from kevmo/homogeneous-16324

DOC: Fix spelling typo - homogenous to homogeneous. (#16324)

mattip committed:
Merge pull request #16337 from seberg/circleci-artifact

CI: Create a link for the circleCI artifact

charris committed:
Merge pull request #16336 from seberg/trivial-refcountfix-add_docstring

BUG: Fix refcounting in add_newdoc

seberg committed:
BUG: Fix dtype leak in `PyArray_FromAny` error path

Also adds a test to bincount which will run into this path.
The leak can be triggered by using a reference count checker
on the test suit (e.g. pytest-leaks).

Closes gh-16339

wrzadkow committed:
MAINT: Remove f-strings in setup.py. (gh-16346)

Remove f-strings from setup.py to allow for an informative error message for python<3.6 users. Closes #16345.

mattip committed:
BUILD: Remove Accelerate support (#15759)

Remove support for Apple Accelerate, since it is buggy. A build error should
occur on most or all setups if linked against Accelerate. Test or import failures
should occur on setups where Accelerate is picked up dynamically.

Co-authored-by: Warren Weckesser <warren.weckesser@gmail.com>

larsoner committed:
BUG: Indentation for docstrings

jbrockmendel committed:
blackify, troubleshoot tests

seberg committed:
Merge pull request #16298 from eric-wieser/fix-np.info

BUG: np.info does not show keyword-only arguments

anirudh2290 committed:
MAINT: Cleanup 'tools/download-wheels.py' (#16329)

* ENH: Create download dir if not present for download-wheels

* TST: Add tests for download-wheels

* MAINT: use exist_ok=True for os.makedirs in download-wheels

Co-authored-by: Ross Barnowski <rossbar@berkeley.edu>

* MAINT: Remove test for download_wheels

* MAINT: Remove directory creation and error if dir not present

* MAINT: Print total files download and remove err for 0 search results

Co-authored-by: Ross Barnowski <rossbar@berkeley.edu>

charris committed:
Merge pull request #16349 from larsoner/dep

BUG: Indentation for docstrings

charris committed:
Merge pull request #16348 from seberg/fromany-leak

BUG: Fix dtype leak in `PyArray_FromAny` error path

mattip committed:
Merge pull request #16266 from jbrockmendel/cyapi

ENH: include dt64/td64 structs and funcs and tests in __init__.pxd.

Qiyu8 committed:
DOC: Reconstruct Testing Guideline. (#16323)

* reconstruct testing doc

* Update doc/TESTS.rst.txt

* DOC: add docstring for numpy.test

Co-authored-by: Ross Barnowski <rossbar@berkeley.edu>
Co-authored-by: Matti Picus <matti.picus@gmail.com>

charris committed:
MAINT: Streamline download-wheels.

- Makes specifying name patterns simpler
- Makes name pattern reject dev versions unless specified.
- Makes progress printing aligned and more concise.

tylerjereddy committed:
Merge pull request #16362 from charris/update-download-wheels

MAINT: Streamline download-wheels.

dependabot-preview[bot] committed:
MAINT: Bump hypothesis from 5.14.0 to 5.15.1

Bumps [hypothesis](https://github.com/HypothesisWorks/hypothesis) from 5.14.0 to 5.15.1.
- [Release notes](https://github.com/HypothesisWorks/hypothesis/releases)
- [Commits](https://github.com/HypothesisWorks/hypothesis/compare/hypothesis-python-5.14.0...hypothesis-python-5.15.1)

Signed-off-by: dependabot-preview[bot] <support@dependabot.com>

charris committed:
Merge pull request #16368 from numpy/dependabot/pip/hypothesis-5.15.1

MAINT: Bump hypothesis from 5.14.0 to 5.15.1

dependabot-preview[bot] committed:
MAINT: Bump pytest-cov from 2.8.1 to 2.9.0

Bumps [pytest-cov](https://github.com/pytest-dev/pytest-cov) from 2.8.1 to 2.9.0.
- [Release notes](https://github.com/pytest-dev/pytest-cov/releases)
- [Changelog](https://github.com/pytest-dev/pytest-cov/blob/master/CHANGELOG.rst)
- [Commits](https://github.com/pytest-dev/pytest-cov/compare/v2.8.1...v2.9.0)

Signed-off-by: dependabot-preview[bot] <support@dependabot.com>

dependabot-preview[bot] committed:
MAINT: Bump cython from 0.29.17 to 0.29.19

Bumps [cython](https://github.com/cython/cython) from 0.29.17 to 0.29.19.
- [Release notes](https://github.com/cython/cython/releases)
- [Changelog](https://github.com/cython/cython/blob/master/CHANGES.rst)
- [Commits](https://github.com/cython/cython/compare/0.29.17...0.29.19)

Signed-off-by: dependabot-preview[bot] <support@dependabot.com>

charris committed:
Merge pull request #16367 from numpy/dependabot/pip/cython-0.29.19

MAINT: Bump cython from 0.29.17 to 0.29.19

charris committed:
Merge pull request #16369 from numpy/dependabot/pip/pytest-cov-2.9.0

MAINT: Bump pytest-cov from 2.8.1 to 2.9.0

melissawm commented on doc/source/dev/howto-docs.rst:
Since we are trying to get help from people who may have never contributed before, it might be better to spell out what "PR" and "issues" are. (Maybe point links to the github documentation or our own repo).

bjnath replied:
Good point. The links will help and I'll add them. 

It turns out though that the links themselves could be more helpful. The issues link takes you to the issues page but doesn't explain what to do once you're there. The PR link takes you to a page that lists the steps, but there's nothing on the page saying "This is how to do a pull request." (When I landed there I thought I'd been taken to the wrong page.)

Newcomers will want to file issues and PRs for code as well as docs, so this page is not the right place to explain them in detail. Maybe we need a "How to fix what's broken" page. This page would link to it (for details on how to do these steps) and it would link to this page (for details of when to do each).

bjnath replied:
These links appear in #16259 (a new version of this PR).

melissawm commented on doc/source/dev/howto-docs.rst:
I feel like this point, raised by @rossbar in another issue, is important. Is there a way we can keep this in the document somehow?

bjnath replied:
Showing usefulness gradations of feedback, from eagerly accepted down to noise, could indeed be helpful, and I didn't see that earlier. I'll incorporate it.

bjnath replied:
> . Changes that modify wording or rearrange phrasing without changing the technical content are discouraged

I paused over this line, because all the time I've been here this is exactly what I've been doing.  Including this very article. 
 
But I know what's intended.

melissawm replied:
This is different, however. It's part of an effort to bring quality to pages which will (hopefully) be read by newcomers and potential contributors, and they're part of larger structural comments. Every effort here is valuable! 

The original note is more about small doing PR's for commas or typos in docstrings or other material, which by itself is not encouraged :)

bjnath replied:
Thanks.

melissawm reviewed (commented):
Overall, I think this looks good! Thanks for working on this, @bjnath !

bjnath committed:
DOC: Response to PR #16259 comments

Links to PR and issue on first reference; spell out PR.

Detail on which word changes won't fly.

bjnath commented:
I pushed commit f55a3d0 on this branch responding to the two comments.

bjnath commented:
Here's the revised page:
![image](https://user-images.githubusercontent.com/6691888/82847293-54fa8a80-9ebb-11ea-960a-8f4a949bb998.png)

Qiyu8 committed:
ENH: Optimize Cpu feature detect in X86, fix for GCC on macOS (#16297)

* Fix cpu detection for X86, GCC, and macOS

* Optimize feature detection

Co-authored-by: Matti Picus <matti.picus@gmail.com>

bjnath committed:
Doc: Further work on #16259

Reduce the negativity in the fixes section. Separate sections
for fixes and new pages. Consolidate advice on new pages.

bjnath commented:
Committed 967a735 which adds wording improvements to f55a3d0.

![image](https://user-images.githubusercontent.com/6691888/82910540-bd3b8180-9f38-11ea-98f1-0e584017abf2.png)

bjnath commented:
I may have messed up this most recent commit -- I see the unrelated 5436551 listed above with my own commits.

rgommers commented:
The typical way this pulling in of unrelated commits happens is by doing a rebase inside an already ongoing rebase. Is that what you could have done by accident?

If there's only a few new commits you want to keep, cherry-picking them into a new branch and then force-pushing back to this PR is probably the best way to clean it up.

charris commented:
> cherry-picking them into a new branch

Merging to a new branch off master seems to work as well, the duplicate commits go away.

bjnath commented:
My git history was tangled beyond my ability to repair; I got a new fork and continued this PR as gh-16385.  That has the same `newdoc-howto.rst` revisions as the latest version here. This PR should be closed without merging.
